signac/core/syncedlist.py

```diff
@@ -0,0 +1,157 @@
+# Copyright (c) 2020 The Regents of the University of Michigan
+# All rights reserved.
+# This software is licensed under the BSD 3-Clause License.
+from collections.abc import Sequence
+from collections.abc import MutableSequence
+
+from .collection_api import SyncedCollection
+from .collection_api import NUMPY
+
+if NUMPY:
+    import numpy
+
+
+class SyncedList(SyncedCollection, MutableSequence):
+    """Implementation of list data structure."""
+
+    base_type = 'non-str Sequence'  # type: ignore
+
+    def __init__(self, data=None, **kwargs):
+        super().__init__(**kwargs)
+        if data is None:
+            self._data = []
+        else:
+            if NUMPY and isinstance(data, numpy.ndarray):
+                data = data.tolist()
+            with self._suspend_sync():
+                self._data = [self.from_base(data=value, parent=self) for value in data]
+            self.sync()
+
+    @classmethod
+    def is_base_type(cls, data):
+        """Checks whether the data is of base type of SyncedDict.
+
+        Parameters
+        ----------
+        data: any
+            Data to be checked
+
+        Returns
+        -------
+        bool
+        """
+        if isinstance(data, Sequence) and not isinstance(data, str):
+            return True
+        elif NUMPY:
+            if isinstance(data, numpy.ndarray):
+                return True
+        return False
+
+    def to_base(self):
+        """Converts the SyncedList object to list.
+
+        Returns
+        -------
+        converted: list
+            List containing the conveted SyncedList object."""
+        converted = list()
+        for value in self._data:
+            if isinstance(value, SyncedCollection):
+                converted.append(value.to_base())
+            else:
+                converted.append(value)
+        return converted
+
+    def _dfs_update(self, data=None):
```

Comment:
I've always personally found the `_dfs` naming confusing. I would prefer to name this `_update`, IMO it's more important to name the function based on what it does rather than how it does it. However, we should absolutely document internally that it uses a depth first algorithm; developer documentation is where algorithmic information belongs I think.

I'm commenting on this in the list implementation rather than the dict implementation, but this change needs to be made in the other functions too. Would it be possible to unify the implementations? Aside from the default type of `data` and the error (both of which can be set using class variables), the other differences don't seem critical to maintain. Could we merge implementations and consider optimizing after, or are there significant performance problems with getting rid of the list-specific optimizations here?

Reply:
I don't think we are using specific optimization in both methods but I don't know how to combine both of them.

Reply:
There are a few separate topics to address in this conversation:
- I looked at the code and I'm not sure how I would combine these implementations for SyncedDict/SyncedList either. I think it is fine to leave it as it is.
- _Copied from a separate comment in my current PR review:_ Please replace "dfs" with "depth-first traversal" in all docstrings. It's not really search, just traversal, and the acronym is not obvious.
- I agree that the "depth-first" detail should be in developer documentation and not necessarily in the method name. The name has been changed to `_update` but the corresponding tests are still named `dfs_update`. I will make suggested changes to those test names in my PR review.

@vyasr Vyas, please weigh in here if you have an idea for combining the `_update` implementations. Otherwise please resolve this conversation.